### PR TITLE
Fix OOM when running stdlib

### LINKED ISSUE #42924: [Bug]: OOM when running stdlib level 8 and 6
### Description

PRs to nutcracker branch fails when trying to run stdlib 8 and 6 due to OOM. Looking at the stack trace most likely culprit is [`insertAtomAtIndex`](https://github.com/ballerina-platform/ballerina-lang/blob/1083bb87aa2516d6e234fd7d9fdb7f6bd0502a4c/semtypes/src/main/java/io/ballerina/types/Env.java#L143C17-L143C34) which is triggered when trying to load BIR. Need to investigate (if that is the actual source of the error) whether this is caused by

1. There is something wrong with how we serialize/de-serialize BIR causing us to fill the array untill we run out of memeory
2. Some change to those standard library (adding more types) is actually creating "too many" types

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
$subject

Fixes #42924

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRTypeWriter.java [hunk 1]
```diff
@@ -131,7 +131,6 @@ public BIRTypeWriter(ByteBuf buff, ConstantPool cp, Env typeEnv) {
     }
 
     public void visitType(BType type) {
-        writeSemType(type.semType());
```
Comment (lochana-chathura) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRTypeWriter.java:134]: Now we don't serialize the semtype? We serialize/deserialize BType and we re-construct the semtype from that deserialized BType right?

**Review by lochana-chathura — COMMENTED** [review]
LGTM.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRTypeWriter.java (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRTypeWriter.java:134]: Mostly yes ([updated](https://github.com/ballerina-platform/ballerina-lang/issues/42924#issuecomment-2229867817) the issue with the rational behind this)

File: semtypes/src/main/java/io/ballerina/types/ComplexSemTypeImpl.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ */
+package io.ballerina.types;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * @param all             all & (1 << c) is non-zero iff this type contains all the basic type with code c
+ * @param some            some & (1 << c) is non-zero iff this type contains some but not all the basic type with code
+ *                        c
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply semtypes/src/main/java/io/ballerina/types/ComplexSemTypeImpl.java]: Fixed

File: semtypes/src/main/java/io/ballerina/types/subtypedata/BddNode.java [hunk 3]
```diff
@@ -11,27 +11,36 @@
  *  Unless required by applicable law or agreed to in writing,
  *  software distributed under the License is distributed on an
  *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
- *  KIND, either express or implied. See the License for the
+ *  KIND, either express or implied.  See the License for the
  *  specific language governing permissions and limitations
  *  under the License.
+ *
  */
+
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply semtypes/src/main/java/io/ballerina/types/subtypedata/BddNode.java]: Fixed

File: semtypes/src/main/java/io/ballerina/types/subtypedata/BddNodeSimple.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
+/*
+ *  Copyright (c) 2024, WSO2 LLC. (http://www.wso2.org).
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply semtypes/src/main/java/io/ballerina/types/subtypedata/BddNodeSimple.java]: Fixed

File: semtypes/src/main/java/io/ballerina/types/subtypedata/BddNodeSimple.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply semtypes/src/main/java/io/ballerina/types/subtypedata/BddNodeSimple.java]: Fixed

File: semtypes/src/main/java/io/ballerina/types/subtypedata/BddNodeImpl.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
+/*
+ *  Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com).
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply semtypes/src/main/java/io/ballerina/types/subtypedata/BddNodeImpl.java]: Fixed

File: semtypes/src/main/java/io/ballerina/types/subtypedata/BddNodeImpl.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import io.ballerina.types.Bdd;
+
+/**
+ * Actual implementation of a generic Bdd node.
+ *
+ * @param atom   the atom that this node represents
+ * @param left   path that include this node's atom positively
+ * @param middle path that doesn't include this node's atom
+ * @param right  path that include this node's atom negatively
+ * @since 2201.8.0
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply semtypes/src/main/java/io/ballerina/types/subtypedata/BddNodeImpl.java]: Fixed

**Review by lochana-chathura — APPROVED** [review]

**Pull request merged**